Introduce visual testing framework for RViz

botteroa-si opened the pull request:
Closes #120.

Possible implementation of a visual testing framework for RViz, in the spirit of what is done by Ogre.

**Usage**
CMake flags are used to enable the visual tests and to take/update the reference images. In particular to run test:  `ament test --cmake-args -DEnableVisualTests=TRUE`. (for more detailed info see the README file).

**Current features**
- launch RViz and interact with the displays panel;
- take screenshot of the main render window;
- compare taken screenshots with existing reference images (which, in turn, can be taken or updated with another CMake flag); 
- convenience class BasePageObject to write page objects for the various displays;
- works on Linux, MacOS and Windows.

To have an idea of how tests are written, see the example tests. 

**Missing features (in progress)**
- convenience method to write and start publishers for display topics;
- possibility to take screenshots of secondary render windows (e.g. for camera or image displays).

**Limitations**
- the comparison of pictures is pixel-wise. This implies that screenshots taken on different machines will generally be regarded as different; 

**Modification of existing RViz code**
Some files in `rviz_common` and `rviz_rendering` have been modified. In particular we had to expose the render window to make it possible to take screenshots, and we had to give a name to some QWidgets so that they could be accessed from the tests.

**Points to be discussed**
- how do we want to expose also secondary render windows?
- how to integrate this with Jenkins;
- how do we want to deal with the difference of images from different screens?

@wjwwood, please let us know what your opinion about this is.

botteroa-si committed:
Expose RenderWindow and loadDisplayConfig() method, make it possible to take screenshot, and give names to various widgets.

botteroa-si committed:
Set up visual testing framework

Fix setting grid offset on Ubuntu

Change display indexes after rebasing on new master.

wjwwood commented:
> how do we want to expose also secondary render windows?

I don't know if I have a suggestion here, I'd actually ask you guys at this point. I assume that we'd need a new API to expose additional render windows, but I don't know exactly what the tests need to actually use them.

> how to integrate this with Jenkins;

Depends on what you're asking, whether you mean enabling it (since it has cmake options to opt in) or having the ability to actually run rviz (needing a GUI).

For enabling it on demand, we can pass special arguments when manually creating a job to enable these tests when we want.

For having CI machines which can actually run rviz, that's an open issue that I don't have an answer for right now. I believe our Windows and macOS machines can run rviz since they're physical machines in our building. The Linux ones are in docker on headless EC2 servers, so not sure what we'll have to do there.

/cc @nuclearsandwich FYI

> how do we want to deal with the difference of images from different screens?

Great question :), I imagine there's a whole field of study here, but I think Ogre has a "fuzzy" image matching, which might get us into trouble but might also be required for being able to realistically run it on different machines.

> @wjwwood, please let us know what your opinion about this is.

In general this looks nice, but we have to figure out how much the pixel perfect issue is going to bite us.

botteroa-si committed:
Enable possibility to take screenshots from secondary render windows.

botteroa-si committed:
Create CameraDisplay and ImageDisplay page objects.

botteroa-si committed:
Give a name to the camera and image displays render windows, to make them accessible.

botteroa-si committed:
Create PageObjectWithWindow class and refactoring

botteroa-si committed:
Extract Executor class to avoid dependencies from VisualTestFixture plus refactoring

botteroa-si committed:
Update comparison method to use MSE index

botteroa-si committed:
Add missing PointCloudPageObject methods and further refactoring

botteroa-si committed:
Add a PointCloud test with publishers

botteroa-si committed:
Update README and documentation

botteroa-si commented:
@wjwwood , we have now updated and polished the visual testing framework, with the addition of the following features:

- there is now the possibility to take and compare screenshots also of secondary render windows (to enable this it is sufficient to give a name to the `rviz_rendering::RenderWindow` member of the corresponding display). This solution has been implemented for the Camera display and for the Image display;

- if the two images are not identical, then the MSE (Mean Square Error) between the two will be calculated (it is a value between 0 and 1), and the test will fail if this index is higher than a threshold value which can be set by the user and pass otherwise. For now the default value of such a threshold is set to 0.01, and, according to our experience, this guarantees good stability across different machines, given that the tests are sensibly written (they should test the presence and correctness of big objects and not concentrate on details);

- an example test is provided, in which the PointCloud display is tested. In this example we also publish PointCloud messages and transforms directly from the test. 
 
- there are now two documentation files: the README, which is a quick start guide that shortly explains how to write and run tests, and a more in depth explanation of how the framework works, what its goal is and how to write stable tests.

We believe that now the framework is ready to be merged.

botteroa-si commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4145)](http://ci.ros2.org/job/ci_linux/4145/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1222)](http://ci.ros2.org/job/ci_linux-aarch64/1222/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3426)](http://ci.ros2.org/job/ci_osx/3426/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4249)](http://ci.ros2.org/job/ci_windows/4249/)

wjwwood commented on rviz_visual_tests/docs/README.md:
```diff
@@ -0,0 +1,89 @@
+
+      /// Compare test screenshots with the reference ones:
+      assertScreenShotsIdentity();
+    }
```
nitpick: the indentation of this code snippet looks weird, either the first line should be indented or the rest should be unindented once.

wjwwood replied:
The example test looks really nice, btw :D

wjwwood commented on rviz_visual_tests/docs/README.md:
```diff
@@ -0,0 +1,89 @@
+# A quick start guide: how to write and run RViz visual tests
+A deeper explanation on what this testing framework is about and on how it works can be found in
+[this document](documentation.md).
```
Please use the "one line per sentence rule", see: https://github.com/ros2/ros2/wiki/Developer-Guide#markdown

It helps keep diff's small, and prevents you from spending lots of time manually wrapping lines. We require wrapping of lines in doc blocks because code linters usually can't differentiate between source code and doc blocks.

wjwwood commented on rviz_visual_tests/docs/documentation.md:
```diff
@@ -0,0 +1,245 @@
+method may be provided.
+
+- If the dimensions of the test image are different from the ones of the reference picture, then 
+the test screenshot will be resized to match the reference one before the comparison is performed. 
```
How is the resizing done? scaling the image, or centering and cropping?

wjwwood commented on rviz_visual_tests/package.xml:
```diff
@@ -0,0 +1,31 @@
+  <description>
+    3D testing framework for RViz.
+  </description>
+  <maintainer email="empty@empty.com">empty</maintainer>
```
This needs to be set. If you guys cannot, please put me there `<maintainer email="william@openrobotics.org">William Woodall</maintainer>`

wjwwood commented on rviz_visual_tests/package.xml:
```diff
@@ -0,0 +1,31 @@
+  <maintainer email="empty@empty.com">empty</maintainer>
+  <license>BSD</license>
+
+  <author></author>
```
Put your name's here or remove it.

wjwwood commented on rviz_visual_tests/package.xml:
```diff
@@ -0,0 +1,31 @@
+<package format="2">
+  <name>rviz_visual_tests</name>
+  <version>0.0.1</version>
```
This needs to match the version of the other packages in this repository, which is `2.0.0`:

https://github.com/ros2/rviz/blob/4aaa893cfc7140d921a73f33236d6cd2df2aca37/rviz2/package.xml#L3

wjwwood commented on rviz_visual_tests/src/internal/image_tester.cpp:
```diff
@@ -0,0 +1,162 @@
+
+#include "image_tester.hpp"
+
+#include <QLocale>  // NOLIT
```
`NOLINT`?

wjwwood reviewed (changes requested):
Generally lgtm, I had a few small requested changes, but not anything that really needs new CI.

One thing that wasn't clear to me is whether or not other packages can define and run additional tests. My guess is that all tests need to be in the `rviz_visual_tests` package.

Also, originally I had imagined that these kinds of tests would be on the "primitives" of `rviz_rendering` first and only later include more "end-to-end" tests which would the require ROS or message types. For example, I was expecting there to be a "point cloud" renderable type in `rviz_rendering`, which could be rendered with different inputs and settings and validated with historical images. Then the fact that in `rviz_common` and `rviz_default_plugins` this "point cloud" primitive was primed with data from the GUI and either `PointCloud` or `PointCloud2` wouldn't make a difference. Though you might want to test that too at some point.

I don't think the outcome is bad because it's different from what I was originally thinking, but I just wanted to mention it. I'm in favor of leaving it as-is, but where the tests occur might be worth considering in the future.

wjwwood commented:
Also, is there a benefit to keeping these tests in their own package instead of putting them in `rviz_rendering_tests`?

botteroa-si committed:
Build as static library and minor refactoring

botteroa-si committed:
Fix maintainer and author

botteroa-si committed:
Rewrite README and documentation with one sentence per line

botteroa-si committed:
Rename package to rviz_visual_testing_framework

botteroa-si committed:
Change implementation of Cmake flags and paths to save images, plus minor refactoring

These changes are needed in order to write visual tests also in other packages.

botteroa-si committed:
Update documentation

botteroa-si commented:
@wjwwood, thank you for your feedback!

We have implemented your requested changes, plus a couple of other additions.
In particular,  we have renamed the package to `rviz_visual_testing_framework` and we now build a library from it. This, together with other changes to the framework structure, allows other packages to define and run visual tests as well (the documentation has also been updated accordingly). 

> Also, is there a benefit to keeping these tests in their own package instead of putting them in rviz_rendering_tests?

Since now it is possible to write and run visual tests also in other packages, our idea would be to leave the code in the `rviz_visual_testing_framework` package, and write the tests in the relevant package (e.g. in `rviz_default_plugins`).

> How is the resizing done? scaling the image, or centering and cropping?

The resizing is done by scaling the image to match the dimensions of the reference one (which could be' bigger or smaller). I have added also this information to the documentation.

> Also, originally I had imagined that these kinds of tests would be on the "primitives" of rviz_rendering first and only later include more "end-to-end" tests which would the require ROS or message types. 

concerning this point, we directly opted for an end-to-end testing approach because such end-to-end tests were not possible up to this point. Writing tests based on Ogre scene graph is also possible without this framework and is already done in the tests of some displays in `rviz_default_plugins`.

botteroa-si commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4184)](http://ci.ros2.org/job/ci_linux/4184/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1252)](http://ci.ros2.org/job/ci_linux-aarch64/1252/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3466)](http://ci.ros2.org/job/ci_osx/3466/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4281)](http://ci.ros2.org/job/ci_windows/4281/)

botteroa-si committed:
Fix Windows warnings

botteroa-si commented:
New CI, after fixing Windows warnings:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=4187)](http://ci.ros2.org/job/ci_linux/4187/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=1254)](http://ci.ros2.org/job/ci_linux-aarch64/1254/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=3475)](http://ci.ros2.org/job/ci_osx/3475/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=4283)](http://ci.ros2.org/job/ci_windows/4283/)

botteroa-si committed:
Update Rviz config file after rebasing on master.
